lib/segment/src/payload_storage/schema_storage.rs

```diff
@@ -21,9 +21,10 @@ impl SchemaStorage {
         key: PayloadKeyTypeRef,
         value: &PayloadType,
     ) -> OperationResult<()> {
-        let mut schema = self.schema.write();
+        let schema = self.schema.upgradable_read();
```

Comment:
`upgradable_read` unfortunately is still exclusive for other `upgradable_read`, so this change does not help completely.
I would add another quick check with just a `read` lock before that

Reply:
refactored to use a Double-checked locking on schema update.